packages/yarnpkg-core/sources/Project.ts

```diff
@@ -2015,6 +2060,172 @@ function applyVirtualResolutionMutations({
 
   for (const workspace of project.workspaces) {
     volatileDescriptors.delete(workspace.anchoredDescriptor.descriptorHash);
-    resolvePeerDependencies(workspace.anchoredLocator, true, false);
+    resolvePeerDependencies(workspace.anchoredLocator, new Map(), {first: true, optional: false});
+  }
+
+  enum WarningType {
+    NotProvided,
+    NotCompatible,
+  }
+
+  type Warning = {
+    type: WarningType.NotProvided,
+    subject: Locator,
+    requested: Ident,
+    requester: Ident,
+    hash: string,
+  } | {
+    type: WarningType.NotCompatible,
+    subject: Locator,
+    requested: Ident,
+    requester: Ident,
+    version: string,
+    hash: string,
+    requirementCount: number,
+  };
+
+  const warnings: Array<Warning> = [];
+
+  for (const [rootHash, dependents] of peerDependencyDependents) {
+    const root = allPackages.get(rootHash);
+    if (typeof root === `undefined`)
+      throw new Error(`Assertion failed: Expected the root to be registered`);
+
+    const devirtualizedRootLocator = structUtils.devirtualizeLocator(root);
+    const devirtualizedRoot = allPackages.get(devirtualizedRootLocator.locatorHash);
+    if (typeof devirtualizedRoot === `undefined`)
+      throw new Error(`Assertion failed: Expected the devirtualized root to be registered`);
```

Comment:
Were you intending to do something with the `devirtualizedRoot`? It's not reported as being unused because of the typeof check.